logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/creating/iterable/contains/creators/impl/InAnyOrderEntriesAssertionCreator.kt

```diff
@@ -68,18 +87,30 @@ class InAnyOrderEntriesAssertionCreator<E : Any, T : IterableLike>(
         )
 
         val featureAssertion = featureFactory(count, DescriptionIterableAssertion.NUMBER_OF_OCCURRENCES)
-        val assertions = mutableListOf<Assertion>(explanatoryGroup, featureAssertion)
-        val groupType = if (searchBehaviour is NotSearchBehaviour) {
-            assertions.add(createHasElementAssertion(list))
-            addEmptyAssertionCreatorLambdaIfNecessary(multiConsumableContainer, assertions, searchCriterion, count)
-            DefaultSummaryAssertionGroupType
+        val assertions = mutableListOf<Assertion>(explanatoryGroup)
+        if (searchBehaviour is NotSearchBehaviour) {
+            val emptyAssertionHint = createEmptyAssertionHintIfNecessary(multiConsumableContainer, searchCriterion, count)
+                ?.also { assertions.add(it) }
+            val mismatches = createIndexAssertions(list) { (_, element) ->
+                allCreatedAssertionsHold(multiConsumableContainer, element, searchCriterion)
+            }
+            assertions.add(createExplanatoryGroupForMismatches(mismatches))
+            val hasNext = multiConsumableContainer.hasNext(::identity).holds()
+            return assertionBuilder.fixedClaimGroup
+                .withListType
+                .withClaim(mismatches.isEmpty() && hasNext && emptyAssertionHint == null)
+                .withDescriptionAndEmptyRepresentation(AN_ELEMENT_WHICH)
+                .let {
+                    if (hasNext) it.withAssertions(assertions)
+                    else it.withAssertion(explanatoryGroup)
```

Comment:
I might be wrong but I think this `if` shouldn't be necessary or rather is wrong. I guess you don't show that the assertionCreator-Lambda was empty in case the iterable does not hasNext. Please check, what happens if one writes (please add a test-case):
```
expect(listOf<int>()).notToContain { }
```

Reply:
I agree, that `if` is not necessary here.

I took a look at what happens with that test case and I think we've found a bug. In that test case, the hint for the empty `assertionCreator` is actually getting added twice, once from `createExplanatoryAssertionGroup` in `containsHelpers`, and once more from `createEmptyAssertionHintIfNecessary`. There's some other strange things happening with the empty assertion hints but I checked and these same things are happening on master, so I'll create another branch to investigate these issues.